### PR TITLE
Generate GIF during training

### PR DESCRIPTION
**Pull request by charleygros**
This PR introduces the possibility to generate a GIF during training.

Frames = model predictions on a randomly selected slice from the validation subdataset.

### CODE REVIEW

**Comment by charleygros** [comment]
We are taking this opportunity to make of  `ivadomed` a "true" CLI:
```
usage: ivadomed -c CONFIG [-g GIF] [-h]

MANDATORY ARGUMENTS:
  -c CONFIG, --config CONFIG
                        Path to configuration file.

OPTIONAL ARGUMENTS:
  -g GIF, --gif GIF     Generates a GIF of during training, one frame per
                        epoch for a given slice. The parameter indicates the
                        number of 2D slices used to generate GIFs, one GIF per
                        slice. A GIF shows predictions of a given slice from
                        the validation sub-dataset. They are saved within the
                        log directory.
  -h, --help            Shows function documentation.
````

**Comment by charleygros** [comment]
Naturally, it opens the door to the discussion: Which params should we move from the config file to the CLI? --> #356 .

**Comment by charleygros** [comment]
For the time being, the command looks like:
```
ivadomed -c ivadomed/config/config.json -g 2
```
which generates two GIFs, saved under `log_directory/gifs`.

**Comment by charleygros** [comment]
I am having an issue with matplotlib when saving the GIF to mp4:
```
MovieWriter imagemagick unavailable; using Pillow instead.
Traceback (most recent call last):
  File "/home/GRAMES.POLYMTL.CA/chgroc/py36Jan/lib/python3.6/site-packages/PIL/Image.py", line 2138, in save
    format = EXTENSION[ext]
KeyError: '.mp4'

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/home/GRAMES.POLYMTL.CA/chgroc/py36Jan/bin/ivadomed", line 11, in <module>
    load_entry_point('ivadomed', 'console_scripts', 'ivadomed')()
  File "/home/GRAMES.POLYMTL.CA/chgroc/ivado-medical-imaging/ivadomed/main.py", line 268, in run_main
    run_command(context=context, n_gif=args.gif)
  File "/home/GRAMES.POLYMTL.CA/chgroc/ivado-medical-imaging/ivadomed/main.py", line 185, in run_command
    debugging=context["debugging"])
  File "/home/GRAMES.POLYMTL.CA/chgroc/ivado-medical-imaging/ivadomed/training.py", line 292, in train
    gif_dict["gif"][i_gif].save(path_gif_out)
  File "/home/GRAMES.POLYMTL.CA/chgroc/ivado-medical-imaging/ivadomed/utils.py", line 1041, in save
    animation.save(filename, writer='ffmpeg', fps=10)
  File "/home/GRAMES.POLYMTL.CA/chgroc/py36Jan/lib/python3.6/site-packages/matplotlib/animation.py", line 1137, in save
    writer.grab_frame(**savefig_kwargs)
  File "/usr/lib/python3.6/contextlib.py", line 88, in __exit__
    next(self.gen)
  File "/home/GRAMES.POLYMTL.CA/chgroc/py36Jan/lib/python3.6/site-packages/matplotlib/animation.py", line 253, in saving
    self.finish()
  File "/home/GRAMES.POLYMTL.CA/chgroc/py36Jan/lib/python3.6/site-packages/matplotlib/animation.py", line 555, in finish
    duration=int(1000 / self.fps), loop=0)
  File "/home/GRAMES.POLYMTL.CA/chgroc/py36Jan/lib/python3.6/site-packages/PIL/Image.py", line 2140, in save
    raise ValueError("unknown file extension: {}".format(ext)) from e
ValueError: unknown file extension: .mp4
```

This is because `imagemagick` is not install on rosenberg.
Potential solution: try other writer, eg ffmpeg.
To try / discuss.

**Comment by charleygros** [comment]
A less exciting alternative would be to save each frame as png instead of a GIF.

**Comment by jcohenadad** [comment]
> This is because imagemagick is not install on rosenberg.
Potential solution: try other writer, eg ffmpeg.
To try / discuss.

why not using neuropoly-made gifmaker? https://pypi.org/project/gifmaker/

**Comment by charleygros** [comment]
@jcohenadad I was hoping not to add a dependency by using matplotlib cf file change: utils.AnimatedGif
But yes it worth trying. Thanks :+1:

**Comment by jcohenadad** [comment]
> I was hoping not to add a dependency by using matplotlib cf file change: utils.AnimatedGif

if the individual images are already created, then gifmaker will be handy (given that matplotlib is a "big factory" which can be troublesome sometimes). Note: you can use [this API](https://github.com/neuropoly/gifmaker/blob/master/gifmaker/gifmaker.py#L16) directly

**Comment by charleygros** [comment]
sub-hamburg01__T1w__10

**Comment by charleygros** [comment]
Note: the epochs are indicated in the bottom right corner in red

File: ivadomed/utils.py [hunk 1]
```diff
@@ -996,3 +997,54 @@ def volume_reconstruction(batch, pred, undo_transforms, smp_idx, volume=None, we
                                           batch['gt_metadata'][smp_idx],
                                           data_type='gt')
     return pred_undo, metadata, last_sample_bool, volume, weight_matrix
+
+
+def overlap_im_seg(img, seg):
+    """Overlap image (background, greyscale) and segmentation (foreground, jet)."""
+    seg_zero, seg_nonzero = np.where(seg<=0.1), np.where(seg>0.1)
```
Comment (charleygros) [inline ivadomed/utils.py:1004]: 0.1 :scream:

File: ivadomed/main.py [hunk 2]
```diff
@@ -20,38 +20,45 @@
...
+    optional_args.add_argument('-g', '--gif', required=False, type=int, default=0,
+                               help='Generates a GIF of during training, one frame per epoch for a given slice.'
+                                    ' The parameter indicates the number of 2D slices used to generate GIFs, one GIF '
+                                    'per slice. A GIF shows predictions of a given slice from the validation '
+                                    'sub-dataset. They are saved within the log directory.')
+    optional_args.add_argument('-h', '--help', action='help', default=argparse.SUPPRESS,
+                               help='Shows function documentation.')
+
+    return parser
+
```
Comment (lrouhier) [inline ivadomed/main.py:42]: Great ! 🎉

File: ivadomed/utils.py (hunk 1 above)
Comment (jcohenadad) [reply ivadomed/utils.py:1004]: 🤣

**Comment by jcohenadad** [comment]
https://github.com/ivadomed/ivadomed/pull/374#issuecomment-662962407 is fascinating!

**Pull request merged**